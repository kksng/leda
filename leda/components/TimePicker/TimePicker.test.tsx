import React from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';

import userEvent from '@testing-library/user-event';
import { TimePicker } from './index';

describe('Check TimePicker snapshots collection', () => {
  test('is TimePicker render right?', () => {
    const { container } = render(<TimePicker name="test" />);

    expect(container.firstChild)
      .toMatchSnapshot();
  });
});
describe('Check TimePicker value set test collection', () => {
  test('is TimePicker render right if value set as String', () => {
    const timeValue = '10:10';
    const {
      getAllByRole,
      getByRole,
    } = render(<TimePicker name="test" value={timeValue} />);

    expect(getByRole('textbox'))
      .toHaveAttribute('value');

    expect(getAllByRole('textbox'))
      .toHaveLength(1);

    expect(getByRole('textbox'))
      .toHaveValue(timeValue);
  });
  test('is TimePicker render right if value set as Date', () => {
    const timeValue = new Date('2020.01.01 10:10');
    const {
      getAllByRole,
      getByRole,
    } = render(<TimePicker name="test" value={timeValue} />);

    expect(getByRole('textbox'))
      .toHaveAttribute('value');

    expect(getAllByRole('textbox'))
      .toHaveLength(1);

    expect(getByRole('textbox'))
      .toHaveValue('10:10');
  });
  test('is TimePicker render right if value set as Null', () => {
    const timeValue = null;
    const {
      getAllByRole,
      getByRole,
    } = render(<TimePicker name="test" value={timeValue} />);

    expect(getByRole('textbox'))
      .toHaveAttribute('value');

    expect(getAllByRole('textbox'))
      .toHaveLength(1);

    expect(getByRole('textbox'))
      .toHaveValue('');
  });
});
describe('Check TimePicker attributes test collection', () => {
  test('is TimePicker placeholder set right?', () => {
    const validPlaceholder = 'text';
    const { getByRole } = render(<TimePicker placeholder={validPlaceholder} />);

    expect(getByRole('textbox'))
      .toHaveAttribute('placeholder');

    expect(getByRole('textbox'))
      .toHaveProperty('placeholder', validPlaceholder);
  });
  test('is TimePicker with isDisabled attributes work right?', () => {
    const onFocus = jest.fn();
    const { container, getByRole } = render(<TimePicker isDisabled value="10:10" onFocus={onFocus} />);
    const input = getByRole('textbox');

    input.focus();

    expect(container.querySelectorAll('.disabled-state'))
      .toHaveLength(1);

    expect(input)
      .toHaveAttribute('disabled');

    expect(onFocus)
      .toHaveBeenCalledTimes(0);
  });
  test('is TimePicker with different date format input work right?', () => {
    const validFormat = 'hh:mm';
    const invalidFormat = 'hh-mm';
    const validValue = '10:10';
    const invalidValue = '10-10';
    const { rerender, getByRole } = render(<TimePicker format={validFormat} value={validValue} />);
    const input = getByRole('textbox');

    expect(input)
      .toHaveValue(validValue);

    rerender(<TimePicker format={invalidFormat} value={invalidValue} />);

    expect(input)
      .toHaveValue(invalidValue);
  });
  test('is TimePicker with timeMax and timeMin attriutes work right?', () => {
    const timeMin = [22, 5];
    const timeMax = [23, 5];
    const { rerender, getByRole, debug } = render(<TimePicker timemin={timeMin} timemax={timeMax} />);
    const input = getByRole('textbox');

    expect(input)
      .toHaveAttribute('timemin', timeMin.join(','));

    expect(input)
      .toHaveAttribute('timemax', timeMax.join(','));

    // BB
    // Больше проверить нельзя. Не работает ограничение
  });
});
describe('Check TimmePicker event listeners test collection', () => {
  test('is TimePicker onBlur event listener work right?', () => {
    const validDate = '10:10';
    const validName = 'test';
    const onBlur = jest.fn();
    const { container } = render(<TimePicker value={validDate} name={validName} onBlur={onBlur} />);
    const input = container.querySelectorAll('input.datepicker-input')[0];

    fireEvent.blur(input);

    expect(input)
      .not
      .toHaveFocus();

    expect(onBlur)
      .toBeCalledTimes(1);

    expect(onBlur)
      .lastCalledWith(expect.objectContaining({
        component: expect.objectContaining({
          name: validName,
          value: validDate,
        }),
      }));
  });
  test('is TimePicker onFocus event listener work right?', () => {
    const validName = 'test';
    const validValue = '10:10';
    const onFocus = jest.fn();
    const { container } = render(<TimePicker name={validName} value={validValue} onFocus={onFocus} />);
    const input = container.querySelectorAll('input.datepicker-input')[0];

    fireEvent.focus(input);

    expect(onFocus)
      .toHaveBeenCalled();

    expect(onFocus)
      .lastCalledWith(expect.objectContaining({
        component: expect.objectContaining({
          name: validName,
          value: validValue,
        }),
      }));
  });
  test('is TimePicker onChange event listener work right?', () => {
    const validFormat = 'hh:mm';
    const validValue = '10:10';
    const validValueWithotComma = '1011';
    const invalidValue = '11:11';
    const validName = 'test';
    const onChange = jest.fn();
    const { container } = render(<TimePicker value={invalidValue} format={validFormat} name={validName} onChange={onChange} />);
    const input = container.querySelectorAll('input.datepicker-input')[0];

    userEvent.type(input, validValue);

    expect(onChange)
      .toBeCalledTimes(validValue.length);

    expect(onChange)
      .lastCalledWith(expect.objectContaining({
        component: expect.objectContaining({
          name: validName,
          value: validValueWithotComma,
        }),
      }));
  });
  test('is TimePicker onPressEnter event listener work right?', () => {
    const validName = 'test';
    const validValue = '10:10';
    const onEnterPress = jest.fn();
    const { container } = render(<TimePicker name={validName} value={validValue} onEnterPress={onEnterPress} />);
    const input = container.querySelectorAll('input.datepicker-input')[0];

    fireEvent.keyDown(input, {
      charCode: 13,
      code: 13,
      key: 'Enter',
      keyCode: 13,
    });

    expect(onEnterPress)
      .toHaveBeenCalled();

    expect(onEnterPress)
      .lastCalledWith(expect.objectContaining({
        component: expect.objectContaining({
          name: validName,
          value: '', // Ошибка в работе компоненте
        }),
      }));
  });
});
