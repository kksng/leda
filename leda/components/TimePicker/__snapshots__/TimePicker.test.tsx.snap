// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check TimePicker snapshots collection is TimePicker render right? 1`] = `
<div
  class="datepicker-wrapper"
>
  <div
    class="datepicker-input-wrapper"
  >
    <input
      aria-invalid="false"
      class="datepicker-input"
      maxlength="6"
      name="test"
      value=""
    />
    <span
      class="datepicker-icons-wrapper"
    />
  </div>
</div>
`;
